nav2_map_server/src/map_saver_server.cpp

```diff
@@ -0,0 +1,163 @@
+// Copyright 2019 Rover Robotics
+// Copyright (c) 2008, Willow Garage, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_map_server/map_saver_server.hpp"
+
+#include <memory>
+#include <string>
+
+namespace nav2_map_server
+{
+MapSaverServer::MapSaverServer()
+: nav2_util::LifecycleNode("map_saver_server")
+{
+  RCLCPP_INFO(get_logger(), "Creating");
+
+  save_map_timeout_ = std::chrono::milliseconds(
+    declare_parameter("save_map_timeout", 5000));
+
+  map_saver_ = std::make_shared<MapSaver>();
+}
+
+MapSaverServer::~MapSaverServer()
+{
+  RCLCPP_INFO(get_logger(), "Destroying");
+}
+
+nav2_util::CallbackReturn
+MapSaverServer::on_configure(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Configuring");
+  // Create SaveMap service callback handle
+  auto save_map_callback= [this](
+    const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+    const std::shared_ptr<nav2_msgs::srv::SaveMap::Request> request,
+    std::shared_ptr<nav2_msgs::srv::SaveMap::Response> response) -> void {
+      RCLCPP_INFO(get_logger(), "Saving map from %s topic to %s file",
+        request->map_topic.c_str(), request->map_url.c_str());
+
+      try {
+        // Node parameters for map saving
+        std::vector<rclcpp::Parameter> parameters;
+        parameters.push_back(rclcpp::Parameter("map_topic", request->map_topic));
+        parameters.push_back(rclcpp::Parameter("output_file_no_ext", request->map_url));
+        parameters.push_back(rclcpp::Parameter("image_format", request->image_format));
+        parameters.push_back(rclcpp::Parameter("threshold_free", request->free_thresh));
+        parameters.push_back(rclcpp::Parameter("threshold_occupied", request->occupied_thresh));
+        parameters.push_back(rclcpp::Parameter("map_mode", request->map_mode));
+
+        // Update parameters for new subscription.
+        // Errors that might appear in set_parameters() to be handled
+        // in catch() section below
+        map_saver_->set_parameters(parameters);
+
+        // Read node parameters and make new subscription based
+        // on them in order to save the map in file
+        if (!map_saver_->setMapSubscription()) {
+          RCLCPP_ERROR(get_logger(), "Failed to setup new map subscription");
+          response->result = false;
+          return;
+        }
+
+        // Get future just before spinning
+        auto future = map_saver_->map_saved_future();
```

Comment:
You shouldn't need futures in this implementation, I don't believe. You can just wait and have some timeout on the topic reception. I think the future stuff was way overkill.

Reply:
As I understood, the promise-future model initially was made to control program flow, rather than for multi-threading. After subscription was created (currently in `MapSaver::setMapSubscription()`) the `map_saver` should be spinned to get the callback. Usual calling `rclcpp::spin(map_saver_)` will move it in infinite loop and `map_saver_server` will never continue its execution. So, it looks like the promise-future model was added to `map_saver` to use it with `spin_until_future_complete()` routine to stop the flow right after MapSaver::mapCallback() or by timeout.

Here I am not clearly know, what else could we use to simplify the flow?

Reply:
Something like:

```
auto sub = rclcpp_node_->create_subscription<...>(...);
while (!timed_out && rclcpp::ok()) {
  spin_some(rclcpp_node_);
  if (got_map_msg || timed_out) {
    break;
  }
}

```

You can spin a future, spin a node, or spin just the stuff in a callback group until empty.

Reply:
Revised and fully re-worked `MapSaver` class. Fixed starting from https://github.com/ros-planning/navigation2/commit/84cff7d8cf43443e23bc4852129791d665f6e5c7#diff-99885e6562ca50edd8cc80db6362d317R166

Reply:
Thanks a bunch for  that. You saved my sanity.